### DISCUSSION TITLE
Discussion on May 31

### DISCUSSION THREAD

**Discussion by crazyboycjr** [phoenix-dataplane/phoenix#119] [Ideas]
1. returns empty_slabs to a certain global pool
2. add a ref_cnt field in ObjectPage
3. change the meaning of two fields inside ShmPtr
4. change SwitchAddressSpace to only increase the reference counting

**Comment by libertyeagle** [comment]
- use delay dealloc instead of reference counting. Using reference counting have some extra overhead we do not want to bear.
- now we can completely get rid of SwithchAddressSpace
- need to make sure all allocated memory regions are dropped (by shutting down salloc engine). For this, we need to ensure that for each user thread, its corresponding salloc engine is shutdown after mrpc engine and rpcadapter.

**Comment by libertyeagle** [comment]
[CoerceUnsized](https://doc.rust-lang.org/std/ops/trait.CoerceUnsized.html#) and [DispatchFromDyn](https://doc.rust-lang.org/std/ops/trait.DispatchFromDyn.html#) cannot be implemented for ShmNonNull, hence our mrpc's Box does not support casting to trait object (e.g., Box<T> as Box<dyn Any>).

Just take a note on it. I don't think this is needed.

**Comment by libertyeagle** [comment]
ShmView should not be !Send and !Sync. When ShmView is dropped or to_owned, it sends the addr of all recv_mr it uses (i.e., all SgList elements). Each recv_mr is uniquely belongs to a user-app thread / rpcadapter engine, it must be reposted by the same rpcadapter engine. So the user thread who receives the message (ShmView) must also be the one to dealloc it.

**Comment by libertyeagle** [comment]
ShmView should have a lifetime less than that of the client/server

**Comment by libertyeagle** [comment]
Consider the case where a user thread can have multiple clients / servers. Garbage collection becomes more compelx in this case.
